docs/docs/genai/prompt-registry/optimize-prompts.mdx

```diff
@@ -169,14 +205,124 @@ def predict(question: str, prompt_uri: str) -> str:
+    def optimize(
+        self,
+        prompt: PromptVersion,
+        target_llm_params: LLMParams,
+        train_data: pd.DataFrame,
+        scorers: list[Scorer],
+        objective: Optional[Callable[[dict[str, Any]], float]] = None,
+        eval_data: Optional[pd.DataFrame] = None,
+    ) -> OptimizerOutput:
+        # Implement your custom optimization logic here
+
+        optimized_template = f"Please answer accurately: {prompt.template}"
+
+        return OptimizerOutput(
+            optimized_prompt=optimized_template,
+            optimizer_name="CustomOptimizer",
+            final_eval_score=0.85,
+            initial_eval_score=0.75,
+        )
+
+
+# Use the custom optimizer
+result = mlflow.genai.optimize_prompt(
+    target_llm_params=LLMParams(model_name="openai/gpt-4o-mini"),
+    prompt=prompt,
+    train_data=train_data,
+    scorers=[exact_match],
+    optimizer_config=OptimizerConfig(algorithm=CustomOptimizer),
+)
+```
+
+### DSPyPromptOptimizer
+
+For DSPy-based optimizations, extend the <APILink fn="mlflow.genai.optimize.DSPyPromptOptimizer" >DSPyPromptOptimizer</APILink> class, which provides DSPy integration infrastructure:
+
+```python
+import dspy
+from typing import Callable
+from mlflow.genai.optimize import (
+    DSPyPromptOptimizer,
+    OptimizerOutput,
+    format_dspy_prompt,
+)
+from mlflow.entities.model_registry import PromptVersion
+
+
+class CustomDSPyOptimizer(DSPyPromptOptimizer):
+    # Inherit the DSPyPromptOptimizer class and implement the `run_optimization` method.
+    def run_optimization(
+        self,
+        prompt: PromptVersion,
+        program: dspy.Module,
+        metric: Callable[[dspy.Example], float],
+        train_data: list,
+        eval_data: list,
+    ) -> OptimizerOutput:
+        # Use DSPy's optimization components with your custom logic
+        # Example using DSPy's BootstrapFewShot optimizer
+        optimizer = dspy.BootstrapFewShot(
+            metric=metric,
+            max_bootstrapped_demos=self.optimizer_config.max_few_show_examples,
+        )
+
+        # Compile the program
+        compiled_program = optimizer.compile(
+            student=program,
+            trainset=train_data,
+        )
+
+        return OptimizerOutput(
+            optimized_prompt=format_dspy_prompt(compiled_program),
+            optimizer_name="BootstrapFewShot",
+        )
+
+
+# Use the custom DSPy optimizer
+result = mlflow.genai.optimize_prompt(
+    target_llm_params=LLMParams(model_name="openai/gpt-4o-mini"),
+    prompt=prompt,
+    train_data=train_data,
+    scorers=[exact_match],
+    optimizer_config=OptimizerConfig(algorithm=CustomDSPyOptimizer),
+)
+```
+
+:::note
+When using custom optimizers, ensure they return an `OptimizerOutput` object with the optimized prompt and evaluation scores. The DSPy-based optimizer provides additional utilities for working with DSPy's ecosystem of optimization techniques.
```

Comment:
The 2nd sentence here seems unrelated to the note callout. It might be better to replace that sentence with one that explains why it's critical to return the OptimizerOutput that contains those attributes.